sources/megaproxylist.go

```diff
@@ -0,0 +1,103 @@
+package sources
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/nfx/slrp/pmux"
+	"github.com/rs/zerolog/log"
+)
+
+func init() {
+	Sources = append(Sources, Source{
+		ID:        65,
+		Homepage:  "https://www.megaproxylist.net",
+		Frequency: 24 * time.Hour,
+		Seed:      true,
+		Feed:      simpleGen(Megaproxylist),
+	})
+}
+
+var megaproxylistUrl = fmt.Sprintf("https://www.megaproxylist.net/download/megaproxylist-csv-%s_SDACH.zip", time.Now().Format("20060102"))
+
+func unzipInMemory(body []byte) ([]byte, error) {
+	zipReader, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file as zip file: %w", err)
+	}
+
+	// Read all the files from zip archive
+	for _, zipFile := range zipReader.File {
+		if zipFile.Name != "megaproxylist.csv" {
+			continue
+		}
+
+		unzippedFileBytes, err := readZipFile(zipFile)
+		if err != nil {
+			return nil, fmt.Errorf("zip: can't read desire file")
+		}
+		return unzippedFileBytes, nil
+
+	}
+	return nil, fmt.Errorf("zip: empty file")
+}
+
+func readZipFile(zf *zip.File) ([]byte, error) {
+	f, err := zf.Open()
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return ioutil.ReadAll(f)
+}
+
+// Scrapes https://www.megaproxylist.net
+func Megaproxylist(ctx context.Context, h *http.Client) (found []pmux.Proxy, err error) {
+	log.Info().Msg("Loading proxy Megaproxy database")
+
+	resp, err := h.Get(megaproxylistUrl)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Body == nil {
+		return nil, fmt.Errorf("empty body")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+	csvData, _ := unzipInMemory(body)
+	r := csv.NewReader(bytes.NewBuffer(csvData))
+	r.Comma = ';'
+	r.TrimLeadingSpace = true
+
+	// trick to skip header
+	if _, err := r.Read(); err != nil {
+		return nil, fmt.Errorf("failed to read header: %w", err)
+	}
+
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+
+		if len(record) != 4 {
+			continue
+		}
+
+		found = append(found,
+			pmux.NewProxy(fmt.Sprintf("%s:%s", record[0], record[1]),
```

Comment:
https://github.com/nfx/slrp/blob/main/pmux/proxy.go#L227-L230

please check if the proxy you're about to add is not 0. if it's 0, then you can try `net.LookupIP`, because it's a host. otherwise skip

Reply:
as it not coming with port, i have handled this in different way

```
func getIpAddr(address string) (string, error) {
	var addr string
	if net.ParseIP(address) == nil {
		addrs, err := net.LookupAddr(address)
		if err != nil {
			return "", fmt.Errorf("Failed to resolve domain %s: %w", address, err)
		}
		addr = addrs[0]
	} else {
		addr = address
	}
	return addr, nil
}
```